File: Solutions/0350- Intersection of Two Arrays II/Solution.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        my_dict = dict()
        answer = []

        if (len(nums1) < len(nums2)):
            for num in nums1:
                if num not in my_dict:
                    my_dict[num] = 1
                else:
                    my_dict[num] += 1

            for num in nums2:
                if num in my_dict:
                    answer.append(num)
                    if my_dict[num] > 1:
                        my_dict[num] -= 1
                    else:
                        del my_dict[num]
        else:
            for num in nums2:
                if num not in my_dict:
                    my_dict[num] = 1
                else:
                    my_dict[num] += 1

            for num in nums1:
                if num in my_dict:
                    answer.append(num)
                    if my_dict[num] > 1:
                        my_dict[num] -= 1
                    else:
                        del my_dict[num]
        
        return answer

File: Solutions/0350- Intersection of Two Arrays II/Solution.java
import java.util.*;

class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        Map<Integer, Integer> map = new HashMap<>();
        ArrayList<Integer> temp = new ArrayList<>();
        if (nums1.length < nums2.length) {
            for (int num : nums1) {
                if (!map.containsKey(num))
                    map.put(num,1);
                else
                    map.put(num,map.get(num) + 1);
            }

            for (int num : nums2) {
                if (map.containsKey(num)) {
                    temp.add(num);
                    if (map.get(num) > 1)
                        map.put(num,map.get(num) - 1);
                    else
                        map.remove(num);
                }
            }
        } else {
            for (int num : nums2) {
                if (!map.containsKey(num))
                    map.put(num,1);
                else
                    map.put(num,map.get(num) + 1);
            }

            for (int num : nums1) {
                if (map.containsKey(num)) {
                    temp.add(num);
                    if (map.get(num) > 1)
                        map.put(num,map.get(num) - 1);
                    else
                        map.remove(num);
                }
            }
        }
        int[] answer = new int[temp.size()];
        for (int i = 0;i < answer.length;i++) {
            answer[i] = temp.get(i);
        }
        return answer;

    }
}